.github/actions/execute-runner/index.js

```diff
@@ -0,0 +1,173 @@
+        return
+    }
+
+    // If there's no cache hit, then save directory to the cache now.
+    if (isHit) {
+        console.log(`Cache for the step with key ${cacheKey} has been hit. Skip saving.`)
+    }
+    else {
+        console.log(`Save cache for the step with key ${cacheKey}.`)
+
+        try {
+            await cache.saveCache([fullPath], cacheKey)
+        } catch (error) {
+            console.warn(
+                `Can not save cache by key ${cacheKey}.
+                It seems that seems that it has been saved somewhere else.\nOriginal message: ${error}`
+            )
+        }
+    }
+
+    // The step can leave a special file 'paths.txt'.
+    // This file contains paths of the tools that are needed to be added to the system's PATH.
+    const pathsFilePath = path.join(fullPath, "paths.txt")
+    if (fs.existsSync(pathsFilePath)) {
+
+        const lineReader = readline.createInterface({
+            input: fs.createReadStream(pathsFilePath)
+        });
+
+        lineReader.on('line', function (line) {
+            console.log(`Add path ${line}.`);
+            core.addPath(line)
+        });
+    }
+}
+
+async function executeRunner() {
+    // The main action function. It does the following:
+    // 1. Get all cache names of the steps of the given runner and then try to load those caches by using that names.
+    // 2. Execute the runner. If there are cache hits that have been done previously, then the runner will reuse them.
+    // 3. If there are steps, which results haven't been found during the step 1, then the results of those
+    //    steps will be cached using their cache names.
+
+    const runnerFQDN = core.getInput("runner-fqdn");
+    const cacheVersionSuffix = core.getInput("cache-version-suffix");
+    const cacheDir = path.resolve(path.join("agent_build_output", "step_cache"));
+    const cacheKeyRunnerPart = core.getInput("cache-key-runner-part")
+    if (!fs.existsSync(cacheDir)) {
+        fs.mkdirSync(cacheDir,{recursive: true})
+    }
+
+    // Get json list with names of all steps which are needed for this runner.
+    const runner_helper_script_path = path.join("agent_build_refactored", "scripts", "runner_helper.py")
+    // Run special github-related helper command which returns names for all steps, which are used in the current
+    // runner.
+    const code = child_process.execFileSync(
+        "python3",
+        [runner_helper_script_path, runnerFQDN, "--get-all-cacheable-steps"]
+    );
+
+    // Read and decode names from json.
+    const json_encoded_step_names = buffer.Buffer.from(code, 'utf8').toString()
+    const step_cache_names = JSON.parse(json_encoded_step_names)
+
+    const cacheHits = {}
+
+    const finalCacheSuffix = `${cacheKeyRunnerPart}-${cacheVersionSuffix}`
+
+    // Run through step names and look if the is any existing cache for them.
+    for (let name of step_cache_names) {
+        console.log(name);
+        cacheHits[name] = await checkAndGetCache(
```

Comment:
Probably want to improve logging a bit here to make it easier to see what is going on - e.g. ``console.log("Processing cache steps"); console.log(`Step name...`).`

Reply:
And we probably also want to log the result of ``checkAndGetCache``. Similar story for line 150 below - should add more logging to make troubleshooting easier (if / when needed).